drivers/spi/spi_sifive.c

```diff
@@ -0,0 +1,319 @@
+	} else {
+		SPI_REG(dev, REG_CSMODE) = SF_CSMODE_HOLD;
+	}
+
+	for (int i = 0; i < send_len; i++) {
+
+		/* Send a frame */
+		if (tx_buf->buf != NULL) {
+			spi_sifive_send(dev, ((u8_t *) tx_buf->buf)[i]);
+		} else {
+			/* Send dummy bytes */
+			spi_sifive_send(dev, 0);
+		}
+
+		/* Receive a frame */
+		if (rx_buf->buf != NULL) {
+			/* TODO: what happens with frame lengths > 8? */
+			((u8_t *) rx_buf->buf)[i] = (u8_t) spi_sifive_recv(dev);
+		} else {
+			/* Discard returned value */
+			spi_sifive_recv(dev);
+		}
+	}
+
+	if (!hw_cs_control) {
+		spi_context_cs_control(&SPI_DATA(dev)->ctx, false);
+	} else {
+		SPI_REG(dev, REG_CSMODE) = SF_CSMODE_AUTO;
+	}
+
+	return 0;
+}
+
+/* API Functions */
+
+int spi_sifive_init(struct device *dev)
+{
+	/* Disable SPI Flash mode */
+	CLR_BIT(SPI_REG(dev, REG_FCTRL), SF_FCTRL_EN);
+
+	/* Make sure the context is unlocked */
+	spi_context_unlock_unconditionally(&SPI_DATA(dev)->ctx);
+	return 0;
+}
+
+int spi_sifive_transceive(struct device *dev,
+		const struct spi_config *config,
+		const struct spi_buf_set *tx_bufs,
+		const struct spi_buf_set *rx_bufs)
+{
+	int rc = 0;
+
+	/* Lock the SPI Context */
+	spi_context_lock(&SPI_DATA(dev)->ctx, false, NULL);
+
+	/* Configure the SPI bus */
+	SPI_DATA(dev)->ctx.config = config;
+
+	/*
+	 * Flag to determine if we want the SPI peripheral hardware to control
+	 * the CS line
+	 */
+	bool hw_cs_control = false;
+
+	/* If the chip select configuration is not present */
+	if (config->cs == NULL) {
+		hw_cs_control = true;
+	}
+
+	if (!hw_cs_control) {
+		/*
+		 * If the user has requested manual GPIO control, ask the
+		 * context for control and disable HW control
+		 */
+		spi_context_cs_configure(&SPI_DATA(dev)->ctx);
+		SPI_REG(dev, REG_CSMODE) = SF_CSMODE_OFF;
+	} else {
+		/*
+		 * Tell the hardware to control the requested CS pin.
+		 * NOTE:
+		 *	For the SPI peripheral, the pin number is not the
+		 *	GPIO pin, but the index into the list of available
+		 *	CS lines for the SPI peripheral.
+		 */
+		SPI_REG(dev, REG_CSID) = config->slave;
+		SPI_REG(dev, REG_CSMODE) = SF_CSMODE_AUTO;
+	}
+
+	rc = spi_config(dev, config->frequency, config->operation);
+	if (rc < 0) {
+		spi_context_release(&SPI_DATA(dev)->ctx, rc);
+		return rc;
+	}
+
+	/* Determine the number of transfers to perform */
+	int num_xfers = 0;
+
+	if (tx_bufs == NULL && rx_bufs == NULL) {
+		/* Nothing to do */
+		spi_context_release(&SPI_DATA(dev)->ctx, 0);
+		return 0;
+	} else if (tx_bufs == NULL) {
+		num_xfers = rx_bufs->count;
+	} else if (rx_bufs == NULL) {
+		num_xfers = tx_bufs->count;
+	} else {
+		num_xfers = MIN(tx_bufs->count, rx_bufs->count);
+	}
+
+	/* Default to transferring dummy bytes/discarding received bytes */
+	const struct spi_buf empty_buf = {
+		.buf = NULL,
+		.len = 0,
+	};
+	const struct spi_buf *tx_buf = &empty_buf;
+	const struct spi_buf *rx_buf = &empty_buf;
+
+	/* Perform transfers */
+	for (int i = 0; i < num_xfers; i++) {
```

Comment:
Why not reusing spi_context helpers? so you would not have to have all this code around num_xfers, etc...